src/libraries/System.IO.Packaging/src/System/IO/Packaging/ZipPackagePartPiece.cs

```diff
@@ -0,0 +1,214 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics.CodeAnalysis;
+using System.IO.Compression;
+
+namespace System.IO.Packaging
+{
+    /// <summary>
+    /// A part piece descriptor, made up of a ZipArchiveEntry and a PieceNameInfo.
+    /// </summary>
+    internal sealed class ZipPackagePartPiece : IComparable<ZipPackagePartPiece>
+    {
+        private const string PartPieceFileExtension = ".piece";
+        private const string PartPieceLastExtension = "].last";
+
+        #region Internal Methods
+        /// <summary>
+        /// Return true and create a ZipPackagePartPiece if the name in the input ZipArchiveEntry parses
+        /// as a piece name.
+        /// </summary>
+        /// <remarks>
+        /// No Uri validation is carried out at this level. All that is checked is valid piece syntax.
+        /// This means that the PrefixName returned as part of the ZipPackagePartPiece will not necessarily
+        /// be a part name. For example, it could be the name of the content type stream.
+        /// </remarks>
+        internal static bool TryParse(ZipArchiveEntry zipArchiveEntry, [NotNullWhen(true)] out ZipPackagePartPiece? partPiece)
+        {
+            if (zipArchiveEntry == null)
+                throw new ArgumentNullException(nameof(zipArchiveEntry));
+
+            partPiece = null;
+
+            bool success = TryParseName(zipArchiveEntry.FullName, out PackUriHelper.ValidatedPartUri? partUri, out string? prefixName, out int pieceNumber, out bool isLastPiece);
+
+            if (success)
+            {
+                partPiece = new ZipPackagePartPiece(zipArchiveEntry, partUri!, prefixName!, pieceNumber, isLastPiece);
+            }
+
+            return success;
+        }
+
+        /// <summary>
+        /// Return true and populate the output parameters if the path parses as a piece name.
+        /// </summary>
+        /// <remarks>
+        /// No Uri validation is carried out at this level. All that is checked is valid piece syntax.
+        /// This means that the output prefix name will not necessarily be a part name. For example,
+        /// it could be the name of the content type stream.
+        /// </remarks>
+        internal static bool TryParseName(string path, [NotNullWhen(true)] out PackUriHelper.ValidatedPartUri? partUri, [NotNullWhen(true)] out string? prefixName, out int pieceNumber, out bool isLastPiece)
+        {
+            bool success = true;
+            int searchPosition = path.Length;
+            // All piece names obey the syntax:
+            //  prefix_name "/" "[" 1*digit "]" [".last"] ".piece"
+            // Work backwards from the end of the full path, extracting and checking this metadata in stages.
+            // Stage 1: extract the file extension of ".piece".
+            int resultPosition = path.LastIndexOf(PartPieceFileExtension, StringComparison.OrdinalIgnoreCase);
```

Comment:
Does the spec mention anything about case sensitivity? Can the extension be `.PIECE`?

Reply:
Yes. ECMA-376 part 2, section 6.2.2.3 says:

> Equivalence of part names shall be determined by ASCII case-insensitive matching. Such matching 
compares a sequence of code points as if all ASCII code points in the range 0x41–0x5A (A–Z) were mapped 
to the corresponding code points in the range 0x61–0x7A (a–z). See Reference [1]

Section 7.2.5.2 says:

> If a logical item is a piece, its name shall have suffixes of the following syntax:
> `SuffixName = "/" "[" PieceNumber "]" [".last"] ".piece"`
> `PieceNumber = "0" | NonZeroDigit [1*Digit]`
> `Digit = "0" | NonZeroDigit`
> `NonZeroDigit = "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9" `
> The prefix of a logical item name is the result of removing a suffix, if any, from the logical item name.
> Equivalence of prefixes, and of suffixes shall be determined by ASCII case-insensitive matching. Logical 
names shall be equivalent if their prefixes and suffixes are equivalent. A physical package shall not contain 
equivalent logical item names. 

We validate case-insensitivity in the `ValidPartPiecesAreParsable` test.